File: lib/roar/crs/src/protocol/http.go
// https://github.com/StormWorld0/storm-framework
// License SMF
// Author zxelzy
package main

import (
	"bytes"
	"io"
	"net/http"
	"time"
)

func executeHTTP(req RequestPacket) ResponsePacket {
	// Setup Timeout
	timeout := time.Duration(req.Timeout * float64(time.Second))
	if timeout == 0 {
		timeout = 5 * time.Second
	}

	client := &http.Client{
		Timeout: timeout,
	}

	// Siapkan Body dan Request
	var bodyReader io.Reader
	if req.Body != "" {
		bodyReader = bytes.NewBufferString(req.Body)
	}

	httpReq, err := http.NewRequest(req.Method, req.URL, bodyReader)
	if err != nil {
		return ResponsePacket{Status: "ERROR", Message: err.Error()}
	}

	// Masukkan Headers jika ada
	for key, val := range req.Headers {
		httpReq.Header.Set(key, val)
	}

	// Tembak ke Target!
	resp, err := client.Do(httpReq)
	if err != nil {
		return ResponsePacket{Status: "ERROR", Message: err.Error()}
	}
	defer resp.Body.Close()

	// Baca Balasan Target
	respBody, _ := io.ReadAll(resp.Body)

	// Kembalikan ke Python dengan format yang rapi
	return ResponsePacket{
		Status: "SUCCESS",
		Data: map[string]interface{}{
			"status_code": resp.StatusCode,
			"body":        string(respBody),
		},
	}
}

File: lib/roar/crs/protocol/http.py
# -- https://github.com/StormWorld0/storm-framework
# -- License SMF
# -- Author zxelzy
import smf

from apps.utility.colors import CC
from ..transport import CRS


def http_request(
    method: str,
    url: str,
    headers: dict = None,
    body: str = "",
    redirect: bool = True,
    rawmode: bool = False,
    timeout: float = 5.0,
    verify: bool = True,
    **kwargs,
) -> dict:
    """Wrapper to send HTTP Request to CRS Engine"""

    packet = {
        "primitive": "HTTP_SEND",
        "method": method.upper(),
        "url": url,
        "headers": headers or {},
        "body": body,
        "redirect": redirect,
        "rawmode": rawmode,
        "timeout": timeout,
        "verify": verify,
    }

    if kwargs:
        smf.printf(f"[!] {CC.YELLOW}Unrecognized parameters dropped =>{CC.RESET}", kwargs)

    return CRS.send(packet)
